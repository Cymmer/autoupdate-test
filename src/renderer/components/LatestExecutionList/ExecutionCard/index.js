import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { determinePercentageColorClass } from 'codechum-app-utils';
import GLOBALS from 'codechum-app-globals';
import styles from './styles.module.scss';

import { Card, Text, Icon, Code, Button } from '../../elements';
import {
  buttonTypes,
  textTypes,
  getSyntaxHighlighterLanguageFromExtension,
} from '../../elements/constants';

import { CodyModal } from '../../modals';

const determineExecuteModeData = (executionMode) => {
  if (executionMode === GLOBALS.EXECUTION_MODES.SUBMIT) {
    return {
      executionModeText: 'Submit',
      executionModeColor: GLOBALS.COLOR_CLASSES.GREEN['300'],
    };
  }
  if (executionMode === GLOBALS.EXECUTION_MODES.TEST) {
    return {
      executionModeText: 'Test',
      executionModeColor: GLOBALS.COLOR_CLASSES.BLUE['300'],
    };
  }

  return {};
};

const ExecutionCard = ({ execution, restoreCode }) => {
  const [isToggled, setIsToggled] = useState(false);
  const [isRestoreConfirmation, setIsRestoreConfirmation] = useState(false);
  const [isRestoring, setIsRestoring] = useState(false);

  const {
    execution_number: executionNumber,
    mode: executionMode,
    run_score: runScore,
    perfect_score: perfectScore,
    source_codes: executionSourceCodes,
    programming_language: executionProgrammingLanguage,
  } = execution;

  const executionScore = parseInt(runScore);

  const [mainCode] = executionSourceCodes;

  const { executionModeText, executionModeColor } =
    determineExecuteModeData(executionMode);

  return (
    <>
      {isRestoreConfirmation && (
        <CodyModal
          actions={[
            {
              text: isRestoring ? 'Restoring...' : 'Restore',
              type: buttonTypes.PRIMARY.BLUE,
              disabled: isRestoring,
              onClick: async () => {
                setIsRestoring(true);
                await restoreCode(
                  executionSourceCodes,
                  executionProgrammingLanguage
                );
                setIsRestoring(false);
                setIsRestoreConfirmation(false);
              },
            },
            {
              text: 'Back',
              type: buttonTypes.TERTIARY,
              disabled: isRestoring,
              onClick: () => setIsRestoreConfirmation(false),
            },
          ]}
          body="Are you sure you want to restore this code? This will remove all your existing code in this language."
          codyType={GLOBALS.CODY_TYPES.THINKING}
          data-test="confirmationRestoreModal"
          handleClose={() => setIsRestoreConfirmation(false)}
          isOpen={isRestoreConfirmation}
          title="Restore Code?"
        />
      )}

      <Card className={styles.ExecutionCard}>
        <button
          className={styles.ExecutionCard_title}
          data-test="cardTitleButton"
          type="button"
          onClick={() => setIsToggled(!isToggled)}
        >
          <div className={styles.ExecutionCard_title_label}>
            <Text
              className={styles.ExecutionCard_title_label_executionNumber}
              data-test="executionNumberText"
              type={textTypes.HEADING.XXXS}
            >
              Execution {executionNumber}
            </Text>
            <Text
              className={styles.ExecutionCard_title_label_mode}
              colorClass={executionModeColor}
              data-test="executionModeText"
              type={textTypes.DATA.SM}
            >
              {executionModeText}
            </Text>
          </div>

          <Text
            colorClass={determinePercentageColorClass(
              executionScore,
              perfectScore
            )}
            data-test="executionScoreTotalText"
            type={textTypes.DATA.MD}
          >
            {executionScore}/{perfectScore}
          </Text>

          <div className={styles.ExecutionCard_title_actions}>
            <Button
              data-test="restoreCodeButton"
              icon="replay"
              type={buttonTypes.TEXT.BLUE}
              onClick={(e) => {
                e.stopPropagation();
                setIsRestoreConfirmation(true);
              }}
            >
              Restore Code
            </Button>
            <Icon
              className={styles.ExecutionCard_title_actions_caret}
              data-test="caret"
              icon={isToggled ? 'expand_less' : 'expand_more'}
            />
          </div>
        </button>

        {isToggled && (
          <Code
            className={styles.ExecutionCard_mainCode}
            data-test="mainCode"
            language={getSyntaxHighlighterLanguageFromExtension(
              mainCode.file_extension
            )}
          >
            {mainCode.code}
          </Code>
        )}
      </Card>
    </>
  );
};

ExecutionCard.propTypes = {
  execution: PropTypes.shape({
    execution_number: PropTypes.number.isRequired,

    // only TEST and SUBMIT are allowed
    mode: PropTypes.oneOf(
      Object.values(GLOBALS.EXECUTION_MODES).filter(
        (mode) => mode !== GLOBALS.EXECUTION_MODES.RUN
      )
    ).isRequired,

    // the score of this execution (string because this is Decimal
    // returned from the server)
    run_score: PropTypes.string.isRequired,

    perfect_score: PropTypes.number.isRequired,
    source_codes: PropTypes.arrayOf(
      PropTypes.shape({
        file_extension: PropTypes.string.isRequired,
        code: PropTypes.string.isRequired,
      })
    ).isRequired,

    programming_language: PropTypes.shape({
      id: PropTypes.number.isRequired,
    }).isRequired,
  }).isRequired,

  // the function handler that will restore the code of this execution
  restoreCode: PropTypes.func.isRequired,
};

export default ExecutionCard;
