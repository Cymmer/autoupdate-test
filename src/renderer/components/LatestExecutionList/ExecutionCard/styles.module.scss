@import '../../../scss/mixins';
@import '../../../scss/variables';

@mixin title {
  display: grid;
  align-items: center;
  grid-template-columns: 1fr 1fr 1fr;
  outline: none;
  border: none;
  background-color: transparent;
  padding: 0;
  width: 100%;

  @content;
}

.ExecutionCard {
  padding: 16px 24px;

  &_title {
    @include title {
      cursor: pointer;

      &:hover,
      &:focus,
      &:active {
        .ExecutionCard_text,
        .ExecutionCard_icon {
          color: $blue-300;
        }
      }
    }

    &_label {
      justify-self: start;

      &_executionNumber {
        display: inline-block;
      }

      &_mode {
        margin-left: 16px;
        display: inline-block;
      }
    }

    &_actions {
      @include flex-center;
      justify-self: end;

      &_caret {
        margin-left: 56px;
        @include theme {
          color: t('textColor-400');
        }
      }
    }
  }

  &_mainCode {
    margin-top: 12px;
  }
}