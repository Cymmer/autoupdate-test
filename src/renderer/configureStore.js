import { createHashHistory } from 'history';
import { routerMiddleware } from 'react-router-redux';
import { applyMiddleware, compose, createStore } from 'redux';
import { persistReducer } from 'redux-persist';
import hardSet from 'redux-persist/lib/stateReconciler/hardSet';
import storage from 'redux-persist/lib/storage';
import rootReducer from './ducks/index';

export const configureStore = () => {
  // create middlewares
  const history = createHashHistory();

  const composeEnhancers =
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
  const middleware = applyMiddleware(routerMiddleware(history));

  // create a persist version of the reducer
  const persistConfig = {
    key: 'codechum-desktop',
    storage,
    stateReconciler: hardSet,
  };
  const persistedReducer = persistReducer(persistConfig, rootReducer);

  // create store
  const store = createStore(persistedReducer, composeEnhancers(middleware));

  return store;
};
