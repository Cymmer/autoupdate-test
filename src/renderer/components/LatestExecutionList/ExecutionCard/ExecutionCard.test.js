import React from 'react';
import { mount, shallow } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import { findByTestAttr } from 'codechum-app-utils';
import GLOBALS from 'codechum-app-globals';
import ExecutionCard from './index';

import { configureStore } from '../../../configureStore';

const EXECUTION = {
  execution_number: 1,
  mode: GLOBALS.EXECUTION_MODES.TEST,
  run_score: '10.000',
  perfect_score: 10,
  source_codes: [
    {
      file_extension: 'c',
      code: 'testcode',
    },
  ],
  programming_language: {
    id: 1,
  },
};

describe('RENDERING - LatestExecutionList > ExecutionCard', () => {
  it('should render with default props', () => {
    const component = shallow(
      <ExecutionCard execution={EXECUTION} restoreCode={() => {}} />
    );
    expect(component).toMatchSnapshot();
  });

  it('should render the correct execution number text', () => {
    let component = mount(
      <ExecutionCard execution={EXECUTION} restoreCode={() => {}} />
    );
    let executionNumberText = findByTestAttr(component, 'executionNumberText');
    expect(executionNumberText.text()).toEqual(
      `Execution ${EXECUTION.execution_number}`
    );

    // just to make sure, let's try again with another execution number
    component = mount(
      <ExecutionCard
        execution={{
          ...EXECUTION,
          execution_number: 2,
        }}
        restoreCode={() => {}}
      />
    );
    executionNumberText = findByTestAttr(component, 'executionNumberText');
    expect(executionNumberText.text()).toEqual('Execution 2');
  });

  it('should render the correct execution mode text', () => {
    // this should be "Test" because the execution is a TEST execution
    let component = mount(
      <ExecutionCard execution={EXECUTION} restoreCode={() => {}} />
    );
    let executionModeText = findByTestAttr(component, 'executionModeText');
    expect(executionModeText.text()).toEqual('Test');

    // this should be "Submit" because the execution is a SUBMIT execution
    component = mount(
      <ExecutionCard
        execution={{
          ...EXECUTION,
          mode: GLOBALS.EXECUTION_MODES.SUBMIT,
        }}
        restoreCode={() => {}}
      />
    );
    executionModeText = findByTestAttr(component, 'executionModeText');
    expect(executionModeText.text()).toEqual('Submit');
  });

  it('should render the correct execution mode text color', () => {
    // this should be GREEN because it is TEST
    let component = mount(
      <ExecutionCard execution={EXECUTION} restoreCode={() => {}} />
    );
    let executionModeText = findByTestAttr(component, 'executionModeText');
    expect(executionModeText.prop('colorClass')).toEqual(
      GLOBALS.COLOR_CLASSES.BLUE['300']
    );

    // this should be BLUE because it is SUBMIT
    component = mount(
      <ExecutionCard
        execution={{
          ...EXECUTION,
          mode: GLOBALS.EXECUTION_MODES.SUBMIT,
        }}
        restoreCode={() => {}}
      />
    );
    executionModeText = findByTestAttr(component, 'executionModeText');
    expect(executionModeText.prop('colorClass')).toEqual(
      GLOBALS.COLOR_CLASSES.GREEN['300']
    );
  });

  it('should render the correct execution score/total text', () => {
    const component = mount(
      <ExecutionCard execution={EXECUTION} restoreCode={() => {}} />
    );

    const executionScoreTotalText = findByTestAttr(
      component,
      'executionScoreTotalText'
    );
    expect(executionScoreTotalText.text()).toEqual(
      `${parseInt(EXECUTION.run_score)}/${parseInt(EXECUTION.perfect_score)}`
    );
  });

  it('should render the correct execution score/total color', () => {
    // this should be GREEN because it is PERFECT
    let component = mount(
      <ExecutionCard execution={EXECUTION} restoreCode={() => {}} />
    );
    let executionScoreTotalText = findByTestAttr(
      component,
      'executionScoreTotalText'
    );
    expect(executionScoreTotalText.prop('colorClass')).toEqual(
      GLOBALS.COLOR_CLASSES.GREEN['300']
    );

    // this should now be BLUE because it is greater than 50%
    component = mount(
      <ExecutionCard
        execution={{
          ...EXECUTION,
          run_score: '6.000',
        }}
        restoreCode={() => {}}
      />
    );
    executionScoreTotalText = findByTestAttr(
      component,
      'executionScoreTotalText'
    );
    expect(executionScoreTotalText.prop('colorClass')).toEqual(
      GLOBALS.COLOR_CLASSES.BLUE['300']
    );

    // this should now be RED because it is less than 50%
    component = mount(
      <ExecutionCard
        execution={{
          ...EXECUTION,
          run_score: '3.000',
        }}
        restoreCode={() => {}}
      />
    );
    executionScoreTotalText = findByTestAttr(
      component,
      'executionScoreTotalText'
    );
    expect(executionScoreTotalText.prop('colorClass')).toEqual(
      GLOBALS.COLOR_CLASSES.RED['300']
    );
  });
});

describe('EVENTS - CourseList > CourseCard > Add', () => {
  it('should be able to toggle the main code', () => {
    const store = configureStore();
    const component = mount(
      <Provider store={store}>
        <ExecutionCard execution={EXECUTION} restoreCode={() => {}} />
      </Provider>
    );
    let caret = findByTestAttr(component, 'caret');
    let mainCode = findByTestAttr(component, 'mainCode');

    // check for the values for the closed state
    expect(caret.prop('icon')).toEqual('expand_more');
    expect(mainCode.exists()).not.toBeTruthy();

    // we simulate the opening of the card
    findByTestAttr(component, 'cardTitleButton').simulate('click');
    component.update();

    // so we now check for the opened state
    caret = findByTestAttr(component, 'caret');
    mainCode = findByTestAttr(component, 'mainCode');
    expect(caret.prop('icon')).toEqual('expand_less');
    expect(mainCode.exists()).toBeTruthy();
  });

  it('should be able to show the restore code confirmation modal', () => {
    const store = configureStore();
    const component = mount(
      <Provider store={store}>
        <ExecutionCard execution={EXECUTION} restoreCode={() => {}} />
      </Provider>
    );

    // initially, it shouldn't exist yet
    let confirmationRestoreModal = findByTestAttr(
      component,
      'confirmationRestoreModal'
    );
    expect(confirmationRestoreModal.exists()).not.toBeTruthy();

    // we click the restore code button that will open the confirmation restore modal
    findByTestAttr(component, 'restoreCodeButton').simulate('click');
    component.update();

    // it should now exist
    confirmationRestoreModal = findByTestAttr(
      component,
      'confirmationRestoreModal'
    );
    expect(confirmationRestoreModal.exists()).toBeTruthy();
  });

  it('should be able to close the restore code confirmation modal', () => {
    const store = configureStore();
    const component = mount(
      <Provider store={store}>
        <ExecutionCard execution={EXECUTION} restoreCode={() => {}} />
      </Provider>
    );

    // initially, it shouldn't exist yet
    let confirmationRestoreModal = findByTestAttr(
      component,
      'confirmationRestoreModal'
    );
    expect(confirmationRestoreModal.exists()).not.toBeTruthy();

    // we click the restore code button that will open the confirmation restore modal
    findByTestAttr(component, 'restoreCodeButton').simulate('click');
    component.update();

    // it should now exist
    confirmationRestoreModal = findByTestAttr(
      component,
      'confirmationRestoreModal'
    );
    expect(confirmationRestoreModal.exists()).toBeTruthy();

    // now, let's close the confirmation modal
    act(() => {
      confirmationRestoreModal.prop('actions')[1].onClick();
    });
    component.update();

    // so it should not exist now
    confirmationRestoreModal = findByTestAttr(
      component,
      'confirmationRestoreModal'
    );
    expect(confirmationRestoreModal.exists()).not.toBeTruthy();
  });

  it('should be able to proceed with the restore code', async () => {
    const restoreCodeFn = jest.fn();

    const store = configureStore();
    const component = mount(
      <Provider store={store}>
        <ExecutionCard execution={EXECUTION} restoreCode={restoreCodeFn} />
      </Provider>
    );

    // initially, it shouldn't exist yet
    let confirmationRestoreModal = findByTestAttr(
      component,
      'confirmationRestoreModal'
    );
    expect(confirmationRestoreModal.exists()).not.toBeTruthy();

    // we click the restore code button that will open the confirmation restore modal
    findByTestAttr(component, 'restoreCodeButton').simulate('click');
    component.update();

    // it should now exist
    confirmationRestoreModal = findByTestAttr(
      component,
      'confirmationRestoreModal'
    );
    expect(confirmationRestoreModal.exists()).toBeTruthy();

    // now, let's actually proceed with the process
    await act(async () => {
      await confirmationRestoreModal.prop('actions')[0].onClick();
    });
    component.update();

    expect(restoreCodeFn).toHaveBeenCalledWith(
      EXECUTION.source_codes,
      EXECUTION.programming_language
    );

    // the modal should be closed as well
    confirmationRestoreModal = findByTestAttr(
      component,
      'confirmationRestoreModal'
    );
    expect(confirmationRestoreModal.exists()).not.toBeTruthy();
  });
});
