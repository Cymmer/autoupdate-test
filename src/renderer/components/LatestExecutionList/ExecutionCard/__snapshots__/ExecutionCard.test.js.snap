// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RENDERING - LatestExecutionList > ExecutionCard should render with default props 1`] = `
<Fragment>
  <Card
    className="ExecutionCard"
    id={null}
    onClick={null}
    onMouseEnter={null}
    onMouseLeave={null}
  >
    <button
      className="ExecutionCard_title"
      data-test="cardTitleButton"
      onClick={[Function]}
      type="button"
    >
      <div
        className="ExecutionCard_title_label"
      >
        <Text
          className="ExecutionCard_title_label_executionNumber"
          colorClass="n___700"
          data-test="executionNumberText"
          element={null}
          id={null}
          type="heading___xxxs"
        >
          Execution 
          1
        </Text>
        <Text
          className="ExecutionCard_title_label_mode"
          colorClass="b___300"
          data-test="executionModeText"
          element={null}
          id={null}
          type="data___sm"
        >
          Test
        </Text>
      </div>
      <Text
        className={null}
        colorClass="g___300"
        data-test="executionScoreTotalText"
        element={null}
        id={null}
        type="data___md"
      >
        10
        /
        10
      </Text>
      <div
        className="ExecutionCard_title_actions"
      >
        <Button
          className={null}
          data-test="restoreCodeButton"
          dataTestId={null}
          disabled={false}
          icon="replay"
          id={null}
          isLoading={false}
          isLocked={false}
          kind="button"
          lockedIconClassName={null}
          onClick={[Function]}
          tabIndex={0}
          type="text___blue"
          wrapperClassName={null}
        >
          Restore Code
        </Button>
        <Icon
          className="ExecutionCard_title_actions_caret"
          data-test="caret"
          icon="expand_more"
          id={null}
          isPulsating={false}
          style={null}
        />
      </div>
    </button>
  </Card>
</Fragment>
`;
